cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJob.java

```diff
@@ -0,0 +1,248 @@
+  private static final String PROGRAM_OPTIONS_FILE_NAME = "program.options.json";
+  private static final String CDAP_CONF_FILE_NAME = "cConf.xml";
+  private static final String APP_SPEC_FILE_NAME = "appSpec.json";
+  private static final Gson GSON =
+    ApplicationSpecificationAdapter.addTypeAdapters(new GsonBuilder())
+      .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
+      .registerTypeAdapter(ProgramOptions.class, new ProgramOptionsCodec()).create();
+
+  @Override
+  public void run(RuntimeJobEnvironment runtimeJobEnv) throws Exception {
+    // Get Program Options
+    ProgramOptions programOptions = readJsonFile(new File(PROGRAM_OPTIONS_FILE_NAME), ProgramOptions.class);
+    ProgramId programId = programOptions.getProgramId();
+
+    // Get App spec
+    ApplicationSpecification appSpec = readJsonFile(new File(APP_SPEC_FILE_NAME), ApplicationSpecification.class);
+    ProgramDescriptor programDescriptor = new ProgramDescriptor(programId, appSpec);
+
+    // Update cConf with provided properties. These properties can be used to set configs for twill runner such as
+    // connection string for discovery.
+    CConfiguration cConf = CConfiguration.create();
+    updateConf(runtimeJobEnv, cConf);
+
+    // Create injector and get program runner
+    Injector injector = Guice.createInjector(createmodule(runtimeJobEnv, cConf));
+    ProgramRunner programRunner = injector.getInstance(ProgramRunnerFactory.class).create(programId.getType());
+
+    // Create and run the program. The program files should be present in current working directory.
+    try (Program program = createProgram(cConf, programRunner, programDescriptor, programOptions)) {
+      CompletableFuture<ProgramController.State> programCompletion = new CompletableFuture<>();
+      ProgramController controller = programRunner.run(program, programOptions);
+      controller.addListener(new AbstractListener() {
+        @Override
+        public void init(ProgramController.State currentState, @Nullable Throwable cause) {
+          switch (currentState) {
+            case ALIVE:
+              alive();
+              break;
+            case COMPLETED:
+              completed();
+              break;
+            case KILLED:
+              killed();
+              break;
+            case ERROR:
+              error(cause);
+              break;
+          }
+        }
+
+        @Override
+        public void completed() {
+          programCompletion.complete(ProgramController.State.COMPLETED);
+        }
+
+        @Override
+        public void killed() {
+          programCompletion.complete(ProgramController.State.KILLED);
+        }
+
+        @Override
+        public void error(Throwable cause) {
+          programCompletion.completeExceptionally(cause);
+        }
+      }, Threads.SAME_THREAD_EXECUTOR);
+
+      // Block on the completion
+      programCompletion.get();
+    } catch (ExecutionException e) {
```

Comment:
Seems like you don't need to catch and rethrow since the `run` method is already having `throws Exception`?